### PR TITLE
warn against implementing Freeze

### PR DESCRIPTION
**Pull request by RalfJung**
As [requested](https://github.com/rust-lang/rust/pull/123184#issuecomment-2028531388) by @workingjubilee

### CODE REVIEW

File: library/core/src/marker.rs [hunk 1]
```diff
@@ -817,6 +817,9 @@ pub trait DiscriminantKind {
 /// This can be used to declare that a constant with a generic type
 /// will not contain interior mutability, and subsequently allow
 /// placing the constant behind references.
+///
+/// This trait is a core part of the language, it just just expressed as a trait in libcore for
+/// convenience. Do *not* implement it for other types.
 #[lang = "freeze"]
 #[unstable(feature = "freeze", issue = "121675")]
```
Comment (Nilstrieb) [inline library/core/src/marker.rs:824]: How about some https://github.com/rust-lang/rust/blob/688c30dc9f8434d63bddb65bd6a4d2258d19717c/compiler/rustc_feature/src/builtin_attrs.rs#L828-L835 instead?

Comment (RalfJung) [reply library/core/src/marker.rs:824]: Well we do have impls just below. What would happen with those?

I think longer-term yes we want `rustc_deny_explicit_impl` but that requires compiler work to turn the existing impls into compiler magic.

Comment (Nilstrieb) [reply library/core/src/marker.rs:826]: Oh, right.

Comment (Nilstrieb) [reply library/core/src/marker.rs:826]: Maybe put this section under a `# Safety` header to make it more clear?

File: library/core/src/marker.rs [hunk 2]
```diff
@@ -817,6 +817,13 @@ pub trait DiscriminantKind {
 /// This can be used to declare that a constant with a generic type
 /// will not contain interior mutability, and subsequently allow
 /// placing the constant behind references.
+///
+/// # Safety
+///
+/// This trait is a core part of the language, it just just expressed as a trait in libcore for
```
Comment (y21) [inline library/core/src/marker.rs:823]:
```suggestion
/// This trait is a core part of the language, it is just expressed as a trait in libcore for
```

**Comment by workingjubilee** [comment]
typo. :^)

@bors r-

**Comment by RalfJung** [comment]
Ah, good catch, thanks!

@bors r=Nilstrieb

**Comment by workingjubilee** [comment]
excellent, thank you everyone! I can't wait to tell someone to stop! :^)

**Pull request merged**